File: leetcode/Binary_Tree_Zigzag_Level_Order_Traversal.java
import java.util.*;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        ArrayList<TreeNode>[] l = (ArrayList<TreeNode>[])new ArrayList[2];
        l[0] = new ArrayList<TreeNode>();
        if (root != null) {
            l[0].add(root);
        }
        for (int i = 0; ; ++i) {
            int a = i % 2;
            int b = 1 - a;
            if (l[a].isEmpty()) break;
            l[b] = new ArrayList<TreeNode>();
            ArrayList<Integer> r = new ArrayList<Integer>();
            ListIterator<TreeNode> li = l[a].listIterator(l[a].size());
            while (li.hasPrevious()) {
                TreeNode t = li.previous();
                r.add(t.val);
                if (i % 2 == 0) {
                    if (t.left != null) l[b].add(t.left);
                    if (t.right != null) l[b].add(t.right);
                } else {
                    if (t.right != null) l[b].add(t.right);
                    if (t.left != null) l[b].add(t.left);
                }
            }
            ret.add(r);
        }
        return ret;
    }
}

File: leetcode/Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        vector<vector<int> > traversal;
        if (root == NULL)
            return traversal;
        
        queue<TreeNode*> node_queue;
        node_queue.push(root);
        queue<int> depth_queue;  // 0-based
        depth_queue.push(0);
        int pre_depth = -1;
        while (!node_queue.empty()) {
            TreeNode *node = node_queue.front();
            node_queue.pop();
            int depth = depth_queue.front();
            depth_queue.pop();
            if (pre_depth < depth) {
                if (pre_depth > 0 && pre_depth % 2 == 1)
                    reverse(traversal[pre_depth].begin(),
                            traversal[pre_depth].end());
                pre_depth = depth;
                traversal.push_back(vector<int>());
            }
            traversal[depth].push_back(node->val);
            if (node->left != NULL) {
                node_queue.push(node->left);
                depth_queue.push(depth + 1);
            }
            if (node->right != NULL) {
                node_queue.push(node->right);
                depth_queue.push(depth + 1);
            }
        }
        // Remember to reverse the last row if needed. Forgot at first.
        if (traversal.size() % 2 == 0)
            reverse(traversal[traversal.size() - 1].begin(),
                    traversal[traversal.size() - 1].end());
        return traversal;
    }
};
